src/printer.js

```diff
@@ -2401,45 +2407,44 @@ function printJSXElement(path, options, print) {
   }
 
   // Group by line, recording if there was a hardline.
-  let childrenGroupedByLine;
-  if (children.length === 1) {
-    if (!forcedBreak && willBreak(children[0])) {
-      forcedBreak = true;
+  let groups = [[]]; // Initialize the first line's group
+  children.forEach((child, i) => {
+    // leading and trailing JSX whitespace don't go into a group
+    if (i === 0 && isJSXWhitespace(child)) {
+      groups.unshift(child);
+      return;
+    } else if (i === children.length - 1 && isJSXWhitespace(child)) {
+      groups.push(child);
+      return;
     }
 
-    // Don't wrap a single child in a group as there is no need, and a
-    // lone JSX whitespace doesn't work otherwise because the group
-    // will be printed in flat mode, and we need to print `{' '}` in
-    // break mode.
-    childrenGroupedByLine = [concat([hardline, children[0]])];
-  } else {
-    // Initialize the first line's group
-    let groups = [[]];
-    children.forEach((child, i) => {
-      let prev = children[i - 1];
-      if (prev && willBreak(prev)) {
-        forcedBreak = true;
-
-        // On a new line, so create a new group and put this element in it.
-        groups.push([child]);
-      } else {
-        // Not on a newline, so add this element to the current group.
-        util.getLast(groups).push(child);
-      }
+    let prev = children[i - 1];
+    if (prev && willBreak(prev)) {
+      forcedBreak = true;
 
-      // Ensure we record hardline of last element.
-      if (!forcedBreak && i === children.length - 1) {
-        if (willBreak(child)) forcedBreak = true;
-      }
-    });
+      // On a new line, so create a new group and put this element in it.
+      groups.push([child]);
+    } else {
+      // Not on a newline, so add this element to the current group.
+      util.getLast(groups).push(child);
+    }
 
-    childrenGroupedByLine = [
-      hardline,
-      // Conditional groups suppress break propagation; we want to output
-      // hard lines without breaking up the entire jsx element.
-      concat(groups.map(contents => conditionalGroup([concat(contents)])))
-    ];
-  }
+    // Ensure we record hardline of last element.
+    if (!forcedBreak && i === children.length - 1) {
+      if (willBreak(child)) forcedBreak = true;
+    }
+  });
+
+  const childrenGroupedByLine = [
+    hardline,
+    // Conditional groups suppress break propagation; we want to output
+    // hard lines without breaking up the entire jsx element.
+    concat(groups.map(contents =>
+      Array.isArray(contents)
+        ? conditionalGroup([concat(contents)])
+        : contents
```

Comment:
I don't fully understand all this code, but if it works, it's fine to me. I don't see anything alarming. Thanks!